### PR TITLE
Added language-detection to use as a service

### PR DESCRIPTION
**Pull request by mcarmonaa**
This PR add a program to perform the language detection. It improves how the language is detected adapted to our case of use.

It takes in a file name and tries to guess the language from it. If the language detection is not succesfull, it prints `NEED_CONTET` to stdout, and waits for read the file content through stdin.

```bash
➜  berserker git:(feature/language-detection) go run language-detection/main.go foo.py
Python%
➜  berserker git:(feature/language-detection) go run language-detection/main.go foo  
NEED_CONTENT#!/usr/bin/env python -tt
Python% 
```
https://github.com/src-d/berserker/issues/21

### CODE REVIEW

**Comment by bzz** [comment]
Great! Now that we have a consensus https://github.com/src-d/berserker/issues/21#issuecomment-317375385 let's add gRPC interface!

https://github.com/src-d/berserker/tree/master/extractor is a good example of gRPC component 
 - may be we could rename `language-detections` to `enry-server`
 - add readme file \w description how to build/run/generate protos (Feel free to re-use build/dir structure from https://github.com/src-d/berserker#extractor )
 - no need for using `glide` as dependency manager here yet (can be added later)
 - add all needed dependencies for proteus to generate .proto from the service/model below
 - add and make use of `-max-message-size` and `-profiler` CLI flags

Here is a proposal for gRPC interface

```go
package enry_server

//proteus:generate
func (s *Service) GetLanguage(r *LanguageDetectionRequest) (*LanguageGuess, error) {
}

// proteus:generate
type LanguageDetectionRequest struct {
	FileName      string
        FileContent   []byte
}

// proteus:generate
type LanguageGuess struct {
	Filename      string 
        Status         //done, need_content, same as https://github.com/bblfsh/sdk/blob/master/protocol/protocol.go#L25
}
```

So the client could send the `LanguageDetectionRequest` with only `FileName` set, and then, in case `LanguageGuess.Status` is `NeedContent` - make one more `LanguageDetectionRequest` with both `FineName` and `FileContent` set.

What do you think?

**Comment by mcarmonaa** [comment]
Now language detection runs as a gRPC service following the last comment. Directory and package were renamed to `enrysrv` as `enry_server` doesn't look so idiomatic in Go.

`enrysrv` binary has server and client modes:
* `enrysrv server`
* `enrysrv client -f foo.py`

To see how it works [README](https://github.com/mcarmonaa/berserker/blob/feature/language-detection/enrysrv/README.md)

File: enrysrv/detection.go
Comment (bzz) [inline enrysrv/detection.go]: May be we should remove it as gRPC is now being used instead

**Comment by bzz** [comment]
Great, thank you for the prompt changes!
Let me test and get back to you with this asap today.

**Comment by bzz** [comment]
@mcarmonaa Looks great, but there is one crucial thing missing, to make it usable from the Scala side. We need a generated .proto files and it's dependency `gogo.proto` in a separate dir inside `berserker` with the correct include paths.

What do you think about:
 - adding `./enrysrv/proto` dir (like [this](https://github.com/src-d/berserker/tree/master/extractor/proto/))

Again https://github.com/src-d/berserker/tree/master/extractor is a good example of what's needed here.

It's a bit counter-intuitive, but it's needed for https://scalapb.github.io/grpc.html to be able to produce client Scala code.

**Comment by mcarmonaa** [comment]
@bzz I see, is it ok to have gogo.proto duplicated in extractor and enrysrv? Maybe we could put all proto files in a separate directory under the project's root directory `berserker/proto` and modify `berserker/normalizer/build.sbt`to point it.

```
...
PB.protoSources in Compile := Seq(file("../proto"))
...
```

Or even we could have the `proto` folder directly under normalizer `berkserker/normalizer/proto`

```
...
PB.protoSources in Compile := Seq(file("proto"))
...
```

**Comment by bzz** [comment]
I believe its t's better to keep it in `enrysrv` as these are protos related only to Enry server.

There is not going to be any duplication as whole `extractor` and `normalizer` components/dirs will be deleted from source tree, right after #20 is merged.

**Comment by bzz** [comment]
After some thinking - one more API call would be quite useful to expos: in addition to existing `GetLanguage`, it would be great to have something like

```go
//proteus:generate
func IsNotCode(req *EnryRequest) *EnryResponse {
  return enry.IsVendor(req.FileName) || enry.IsDotFile(req.FileName) || enry.IsDocumentation(req.FileName) || enry.IsConfiguration(req.FileName)
}
```

This call would never get full content of the file in request and would serve as a shortcut for the clients, to 
 - skip actual language detection (may require reading a file)
 - skip UAST parsing by bblfsh (always require reading a file)
for files that are docs, configs or are vendored.

Or is it something that is already handled by current implementation of `getLanguagesByFilenameStrategies`?  If so - what would the client get in such case?

**Comment by mcarmonaa** [comment]
getLanguagesByFilenameStrategies it's already checking out it and returning `need content`in such case. It doesn't seem the best solution.

The idea of `enrysrv` having another call to check for `Documentation/Vendor/DotFiles/Configuration` implies that it would be necessary at least two requests to `enrysrv`.

If you have to check for those things so frequently, maybe it's better that the current rpc returns a status `IgnoredCase` specifying as a language `Vendor/Documentation/...`, and add a field to the request to force detection if you really want to force it.

```
// proteus:generate
type EnryRequest struct {
        FileName    string
        FileContent []byte
        ForceDetection bool
}
```

**Comment by bzz** [comment]
I like your idea about adding extra status. 

1. *New status for skipped/ignored resourced*
To keep first iteration simple, do you you think we could change `NEED_CONTENT` response status to `IsIgnored` for now for `Documentation/Vendor/DotFiles/Configuration`?

    We can always improve API later on, but in this case we really do not need a content, as most probably client will want to skip such files.

Except for that - all other changes look good to me. 

One more thing though - is there a logging in *server side* now?

2. *Logging*
Something similar to `bblfsh server`, that in `info` log level
     - prints to console 1 line for each request
     - has response summary and 
     - and a time it took enry to make a decision
     - (optional) a short name of the strategies, that it took to made a decision  
would really be super-useful for debugging.


3. *Error reporting*
    In many cases, while detecting language by content I get 
    ```
    17/07/27 09:24:17 INFO EnryService: Detecting lang for opt/cmu/registry.cfg
    17/07/27 09:24:17 INFO EnryService: Detected filename: opt/cmu/registry.cfg, lang: , status: NEED_CONTENT
    17/07/27 09:24:17 INFO EnryService: Detecting lang for opt/cmu/registry.cfg by content
    17/07/27 09:24:17 INFO EnryService: Detected filename: opt/cmu/registry.cfg, lang: , status: ERROR
    ```

    It's hard to say: is this an error due to none-utf-8 file content encoding? Or does it simply mean that Enry can not detect language?

    I guess the only way to tell would be by looking at server logs. 
    But may be it's time to add something similar to Bblfsh - [`[]Errors` to response](https://github.com/bblfsh/sdk/blob/master/protocol/protocol.go#L131)

**Comment by mcarmonaa** [comment]
About *Error reporting*:

Enry doesn't take care about encoding, in fact I think it treats all as utf-8. So the only error is reported right now is because enry couldn't detect anything.

We could add `[]Errors` now for possible changes in the future, but the only error replied at the moment would be `couldn't detect language.

**Comment by bzz** [comment]
Got it, thanks for explaining! Let's not add it now, and leave `[]Error` reporting for the future improvements!

**Comment by bzz** [comment]
@mcarmonaa could you please post a comment like `All feedback addressed at <SHA>` so it would be easier to understand what part is ready, etc?

glide install fails for me on this branch, presumably due to https://github.com/src-d/go-errors/releases/tag/v1.0.0 so we should move to using  go-errors.v1 everywhere

```
[INFO]	--> Setting version for gopkg.in/toqueteos/substring.v1 to c5f61671513240ddf5563635cc4a90e9f3ae4710.
[ERROR]	Failed to set version on gopkg.in/src-d/go-errors.v0 to 8bbbeeb767dfdd053b9b45d5a16a4f4ce2c6f694: Unable to update checked out version
[ERROR]	Failed to set references: Unable to update checked out version (Skip to cleanup)
```

**Comment by mcarmonaa** [comment]
@bzz All feedback addressed at commit `changes to use go-errors.v1`

To summarize at this point:
* `enrysrv server` run as a gRPC service. The command has `--max-message-size`, `--loglevel` and `--profiler flags`. It does logging and only returns an error if it can't detect any language.

* The response's status are: `Ok`, `NeedContent`, `Ignored`, `Error`
* `NeedContent`status is returned when the request only carries a filename and it is not enough to detect the language.
* `Ignored` status is returned when the filename is `Documentation/Vendor/DotFiles/Configuration`
* For `enrysrv client` is mandatory to use `-f` flag to take in a filename. If the content is required `-c` flag must be set.
* Use of `go-errros.v1`
* Proto files are under berserker/enrysrv/proto . To generate them just run `go generate`
* To build the binary just run `make build`
* To download dependenciet: `glide install`

[README](https://github.com/mcarmonaa/berserker/blob/feature/language-detection/enrysrv/README.md)

**Comment by bzz** [comment]
@mcarmonaa tested today and works like a charm! Thank you

Please, let me do one more pass and if everything is fine - let's merge it on Monday, if that is ok with you.

**Comment by bzz** [comment]
One thing: current `./proto/` dir structure is

```
proto
├── github.com
│   └── src-d
│       └── berserker
│           └── enrysrv
│               └── generated.proto
└── gogo
    └── protobuf
        └── gogoproto
            └── gogo.proto
```

But if you look in `generated.proto` imports, you will see `import "github.com/gogo/protobuf/gogoproto/gogo.proto";`

Could  you please: re-structure `./proto` in the same way as https://github.com/src-d/berserker/tree/master/extractor/proto ?

File: enrysrv/README.md
Comment (bzz) [inline enrysrv/README.md]: Logs are nice initially to communicate the current state of implementation, but they quickly become distracting and make REAMDE long. 

Do you think we could keep only run commands here and below?

File: enrysrv/cli/common.go
Comment (bzz) [inline enrysrv/cli/common.go:51]: This code looks tremendously similar to https://github.com/bblfsh/server/blob/cf9cc98d152c4e4918a44f31862597fe338e00ab/cmd/bblfsh/common.go#L39

Babelfish server is a project under GPL-3 https://github.com/bblfsh/server/blob/master/LICENSE and Berserker does not have a license yet #12 so, it's proprietary src-d code for now.

You seems not to be the copyright holder (original author) of these lines in Babelfhish, so the only way to **legally** copy-paste somebody's else GPL code - is to paste it to GPL project, as it's copyleft.

So, to merge this PR, one for the 2 things must happen:
 - either with @mcuadros  help we decide Berserker to become GPL3 in #12 
 - or you need to re-write  this code, without copying from GPL3 project

Just for the future, to simplify such cases - copyright can be quite tricky to get right, so in case when on needs to copy a code from somewhere, it's always [better to keep records i.e by saying so in comments](https://github.com/src-d/berserker/pull/22/files#diff-2b40fe2809519a9005b1799fea36042aR1). It simplifies the review and helps in cases, if the license of the original code requires to keep  author's attributions.

**Review by bzz — CHANGES_REQUESTED** [review]

File: enrysrv/README.md
Comment (mcarmonaa) [reply enrysrv/README.md]: Do you mean keep incoming requests logged at INFO level and the rest of the logged information at DEBUG level? So we can launch the service at INFO level by default

File: enrysrv/cli/main.go
Comment (bzz) [inline enrysrv/cli/main.go:17]: Just a nitpick probably and does not block the merging, but would not it be nice to replace this + linker flags in Makefile with `go generate`, what do you think?  Like in https://github.com/a-h/version/blob/master/get_version_linux.sh

File: enrysrv/cli/common.go
Comment (mcarmonaa) [reply enrysrv/cli/common.go:51]: Should I rewrite `parseMaxMessageSize` then?

File: enrysrv/README.md
Comment (mcarmonaa) [inline enrysrv/README.md]: Do you mean keep incoming requests logged at INFO level and the rest of the logged information at DEBUG level? So we can launch the service at INFO level by default

File: enrysrv/cli/common.go
Comment (mcarmonaa) [inline enrysrv/cli/common.go]: Should I rewrite parseMaxMessageSize then?

Comment (bzz) [reply enrysrv/cli/common.go:51]: Is it only one function? I think we can wait for #12 to be resolved and see if Berserker becomes GPL3 project, and that would make code sharing \w Babelfish totally fine.

File: enrysrv/cli/main.go
Comment (mcarmonaa) [reply enrysrv/cli/main.go:17]: I think it's better just link this variables during building time.

Eventually if this project it's integrated with src-d/ci, link these variables trough building flags it's the way [Makefile's CI](https://github.com/src-d/ci/blob/master/Makefile.main#L127) do it.

File: enrysrv/README.md
Comment (bzz) [reply enrysrv/README.md]: No, I think current logging is very good! 
Sorry for not explaining it better - what I meant was, update only README.md file, removing the logging information from it and keeping everything else the same.

File: enrysrv/cli/common.go
Comment (smola) [reply enrysrv/cli/common.go:51]: @bzz We are all src-d employees and code we contribute to these repositories have whatever copyright the company chooses.

Comment (smola) [reply enrysrv/cli/common.go:51]: In other words: we can relicense this particular code with whatever license we want for Berserker. There is no need to license as GPLv3, since we can relicense our own code on-demand.

Comment (bzz) [reply enrysrv/cli/common.go:51]: Good point by @smola. Thank you!

Let's ask original author @abeaumont for explicit approval of inclusion of this code in Berserker, and then src-d as copyright owner could license it later at any terms.

**Comment by bzz** [comment]
Looks good to me, thank you @mcarmonaa ! Great job.

If there are no other discussion/objections - let's merge as soon as we get @abeaumont / @mcuadros  approval on including pice of GPL3 bblfsh code in https://github.com/src-d/berserker/pull/22#discussion_r130305439 above.

**Comment by bzz** [comment]
Merging as licensing was wrapped up by https://github.com/src-d/minutes/pull/35/files#diff-89b474f52e6f6b530e7dca2e343dcb64R31

**Pull request merged**
